hal/src/nRF52840/ble_hal.cpp

```diff
@@ -1289,8 +1303,11 @@ int BleObject::Observer::stopScanning() {
     }
     int ret = sd_ble_gap_scan_stop();
     CHECK_NRF_RETURN(ret, nrf_system_error(ret));
-    if (scanSemaphore_) {
-        os_semaphore_give(scanSemaphore_, false);
+    ATOMIC_BLOCK() {
+        if (isScanning_) {
+            isScanning_ = false;
+            os_semaphore_give(scanSemaphore_, false);
```

Comment:
AFAIK，both `sd_ble_gap_scan_stop()` and `sd_ble_gap_connect_cancel()` wouldn't generate event after being called. The reason why I add `ATOMIC_BLOCK()` here is that when `sd_ble_gap_scan_stop()` is executing it can firstly reset the state, even if a scanning timeout event is generated after passing the checking `if (isScanning_)` on enter this function. Otherwise, this semaphore will be given twice, which would mess up the next scanning procedure.

Reply:
I understand why we are doing this, but still, let's not use synchronization primitives under disabled interrupts. I don't giving it twice will cause problems either. The semaphore is binary, unless it's taken, it cannot be given.

Reply:
Even if it is a binary semaphore, if there is a thread calling `hal_ble_gap_start_scan()` has higher priority than another thread calling `hal_ble_gap_stop_scan()`, once the semaphore is given in ISR and return, the semaphore is taken prior to giving the semaphore here. Then next time calling `hal_ble_gap_start_scan()` will return without any scanned result.

Reply:
Wouldn't something like this work?
```cpp
    bool give = false;
    ATOMIC_BLOCK() {
        if (isScanning_) {
            isScanning_ = false;
            give = true;
        }
    }
    if (give) {
         os_semaphore_give(scanSemaphore_, false);
    }
```